### ISSUE TITLE
Messages are read multiple times when announcing notifications

### ISSUE THREAD

**Issue by rsammelson** [TelegramMessenger/Telegram-iOS#835]
### Description
When 'announce notifications' is turned on for Telegram, each time a new message is received from a recipient all unread messages are announced again before finally announcing the new message.

### Expected Behavior
Only new messages should be announced.

### Steps to Reproduce
1. Turn on announce notifications
2. Receive multiple messages from someone

### Environment
**Device:** `iPhone 13`

**iOS version**: `15.6`

**App version:** `8.9.1`

**Comment by haasn** [comment]
Bug still affects iOS 16.0.

**Comment by flubber89** [comment]
Bug still affects iOS 16.3 & CarPlay.

**Comment by matteogus** [comment]
Bug still present on iOS 16.5.1 and Telegram 9.6.3

**Comment by feljx** [comment]
This has been the most annoying bug on my phone for years now. Please fix.

**Comment by matteogus** [comment]
Will anyone in Telegram look at that?
It’s almost a year that this issue is open.

**Comment by matteogus** [comment]
After upgrading to iPhone 15 Pro, iOS 17.0.3 and Telegram 10.1.2 the issue is still there.

**Comment by Piero24** [comment]
Same problem with IOS 17.2 on iPhone 15 Pro. This problem is really annoying. does anyone have any news?

**Comment by plyusnin** [comment]
This bug doesn’t seem to be well discussed in the internet. I wonder if the other people don’t have it or just not use the announcement function?

**Comment by matteogus** [comment]
Updated to iOS 17.3 and Telegram 10.7.1 and still the same problem, this is very annoying 😡

**Comment by kevindb** [comment]
I'm still seeing (hearing) this issue on iOS 17.4 and Telegram 10.9.2.

**Comment by matteogus** [comment]
We are now at iOS 17.5 and Telegram 10.12, but the problem is still here.
This is ridiculous, even WhatsApp can manage it correctly for years…

**Comment by KihtrakRaknas** [comment]
Issue persists on Telegram iOS 10.14.1.

**Comment by matteogus** [comment]
> Issue persists on Telegram iOS 10.14.1.

Yeah, embarrassing…

**Comment by matteogus** [comment]
iOS 18.1 + Telegram 11.2.1 and this issue persists…

**Comment by alferezkururu966** [comment]
More than 2 years and 4 major ios versions and the issue persist. Amazing tbh.

**Comment by Piero24** [comment]
I totally agree with @matteogus and @alferezkururu966 ... ​​After 2 years not even a single comment about it is absurd... @laktyushin Could you at least give us some information about it?

**Comment by maefiosii** [comment]
iphone 16 with latest software, same problem

**Comment by matteogus** [comment]
I completely lost the hope for a fix.
This is very pathetic…

**Comment by matteogus** [comment]
I think that the problem is with the [INSearchForMessagesIntent](https://developer.apple.com/documentation/intents/insearchformessagesintent) library.
Can anyone, more expert than me, verify that this implementation is ok?

**Comment by TylerEich** [comment]
I'm not a Swift programmer, but I spent a few minutes looking at the code. Based on @matteogus great suggestion, I found [a block of code](https://github.com/TelegramMessenger/Telegram-iOS/blob/0cd1770a0aee75346d91f6eaabe9502627ff4399/Telegram/SiriIntents/IntentHandler.swift#L540-L544) that handles `INSearchForMessagesIntent`.

```swift
if let identifiers = intent.identifiers, !identifiers.isEmpty {
    messages = getMessages(account: account, ids: identifiers.compactMap(MessageId.init(string:)))
} else {
    messages = unreadMessages(account: account)
}
```

It seems that, when no identifiers are passed, all unread messages are retrieved. I wonder if there is some other criteria that could be checked? Hard to know without access to logs or breakpoints.

I wish success to a more capable developer! I have "Announce Notifications" turned off for Telegram and I would love to turn it back on once this is resolved! Thanks for all you do!

**Comment by imorda** [comment]
That's crazy on how long such a simple and annoying issue can be left unattended by telegram devs

**Comment by marcodallagatta** [comment]
Is there any roadmap to work on this? It's insane this has been an issue for three years.

**Comment by matteogus** [comment]
3 years and still counting…
iOS 26.0.1 - Telegram 12.1.1

**Comment by matteogus** [comment]
iOS 26.1 - Telegram 12.2
Still waiting for a bug fix…

**Comment by DannieBGoode** [comment]
This is still happening, it's very annoying

**Comment by Piero24** [comment]
Me waiting for a fix since 3+ years.

![image](https://github.com/user-attachments/assets/29c0c934-2c9e-4f2b-95de-b81bb9400edd)

**Comment by V3lvetStorm** [comment]
I have it too...
iPhone 17 Pro Max - iOS 26.3
Telegram Version: 12.4.1


Waiting a fix in February 2026

**Comment by luponata** [comment]
Still an issue...
